Add an entry to `/etc/hosts` for the current hostname

zmrow opened the pull request:
**Issue number:**
Fixes #1700


**Description of changes:**
```
    schnauzer: Add a new template helper: `add_unresolvable_hostname`
    
    This change adds a new template helper `add_unresolvable_hostname` that
    will attempt to resolve the hostname given to it in the template (from
    `settings.network.hostname`).  If the hostname is unresolvable or
    resolves to localhost (127.0.0.1 or ::1), the helper adds an alias to
    the hostname for both the IPv4 and IPv6 addresses
```
```
    Add a hostname entry to `/etc/hosts` if hostname is unresolvable
    
    Previously, `/etc/hosts` was a static file (in the `release` package)
    that we copied into the image when building Bottlerocket.  In order to
    have an entry in the file corresponding to hostname, a new "hosts"
    service was added to the affected services for
    `settings.network.hostname`.  This new "hosts" service has a templated
    configuration file that uses `settings.network.hostname` to populate the
    entry in `/etc/hosts` if the current hostname is unresolvable in DNS.
    The templated configuration file uses the previously added
    `add_unresolvable_hostname` template renderer helper.
```
```
    migrations: Add migrations for `/etc/hosts` "service"
    
    This change includes the 2 migrations necessary to make `/etc/hosts` an
    "affected service" of `settings.network.hostname`: one to add the
    service and configuration file for the hosts service, and another to
    add the hosts service to `settings.network.hostname`.
```

**Testing done:**
Boot an `aws-k8s-1.19` host and observe proper behavior, run a pod, etc.  `/etc/hosts` looks like:
```
bash-5.0# cat /etc/hosts 
127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4
::1 localhost localhost.localdomain localhost6 localhost6.localdomain6
```
Upgrade the same host to a version including these changes, observe proper boot and behavior.  `/etc/hosts` looks the same because the hostname resolves in EC2:
```
bash-5.0# cat /etc/hosts 
127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4l
::1 localhost localhost.localdomain localhost6 localhost6.localdomain6

```

I followed the same procedure with a VMware variant.  `/etc/hosts` before the upgrade:
```
bash-5.0# cat /etc/hosts 
127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4
::1 localhost localhost.localdomain localhost6 localhost6.localdomain6
```
`/etc/hosts` after the upgrade (`hosts-test` is the hostname of my VM):
```
bash-5.0# cat /etc/hosts 
127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4
::1 localhost localhost.localdomain localhost6 localhost6.localdomain6
127.0.0.1 hosts-test
::1 hosts-test
```




**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

zmrow committed:
migrations: Add migrations for `/etc/hosts` "service"

This change includes the 2 migrations necessary to make `/etc/hosts` an
"affected service" of `settings.network.hostname`: one to add the
service and configuration file for the hosts "service", and another to
add the hosts service to `settings.network.hostname`.

bcressey commented:
I'd prefer that we scope this more narrowly so that we only add the hosts entry if our configured hostname isn't resolvable through DNS.

That's really the only case where we know it's needed and helpful.

zmrow committed:
schnauzer: Add a new template helper: `add_unresolvable_hostname`

This change adds a new template helper `add_unresolvable_hostname` that
will attempt to resolve the hostname given to it in the template (from
`settings.network.hostname`).  If the hostname is unresolvable or
resolves to localhost (127.0.0.1 or ::1), the helper adds an alias to
the hostname for both the IPv4 and IPv6 addresses.

zmrow committed:
Add a hostname entry to `/etc/hosts` if hostname is unresolvable

Previously, `/etc/hosts` was a static file (in the `release` package)
that we copied into the image when building Bottlerocket.  In order to
have an entry in the file corresponding to hostname, a new "hosts"
service was added to the affected services for
`settings.network.hostname`.  This new "hosts" service has a templated
configuration file that uses `settings.network.hostname` to populate the
entry in `/etc/hosts` if the current hostname is unresolvable in DNS.
The templated configuration file uses the previously added
`add_unresolvable_hostname` template renderer helper.

zmrow commented:
^ Addresses @bcressey 's concern and adds a new template renderer helper that will only add the entry to `/etc/hosts` if the hostname isn't resolvable or resolves to localhost.

zmrow commented:
^ Splits the template helper into its own commit

zmrow commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -935,8 +1022,9 @@ fn pause_registry<S: AsRef<str>>(region: S) -> String {
 /// Calculates and returns the amount of CPU to reserve
-fn kube_cpu_helper(num_cores: usize) -> Result<String, TemplateHelperError>{
-    let num_cores = u16::try_from(num_cores).context(error::ConvertUsizeToU16 { number: num_cores })?;
+fn kube_cpu_helper(num_cores: usize) -> Result<String, TemplateHelperError> {
```
This was a `rustfmt` change

zmrow commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -949,7 +1037,11 @@ fn kube_cpu_helper(num_cores: usize) -> Result<String, TemplateHelperError>{
         }
     };
-    Ok(format!("{}{}", cpu_to_reserve.floor().to_string(), millicores_unit))
+    Ok(format!(
```
Same here - thanks `rustfmt`

zmrow commented:
^ rebased on `develop`

webern commented on sources/Cargo.toml:
```diff
@@ -24,8 +24,8 @@ members = [
     "api/migration/migration-helpers",
     "api/shibaken",
 
-    # "api/migration/migrations/vX.Y.Z/...
-    # (all migrations currently archived; replace this line with new ones)
```
Maybe leave some comment here identifying this as the migration section?

zmrow replied:
I'll add back the `# "api/migration/migrations/vX.Y.Z/...` comment since that's what was previously there.

webern commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1607,3 +1699,50 @@ mod test_kube_cpu_helper {
+    fn resolves_to_localhost_renders_entries() {
+        let result = setup_and_render_template(
+            "{{add_unresolvable_hostname name}}",
+            &json!({"name": "localhost"}),
```
I guess this test should always work on every system?

zmrow replied:
I believe it should?  I'm willing to be told otherwise though.

webern commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -1607,3 +1699,50 @@ mod test_kube_cpu_helper {
+    fn resolvable_hostname_renders_nothing() {
+        let result = setup_and_render_template(
+            "{{add_unresolvable_hostname name}}",
+            &json!({"name": "amazon.com"}),
```
Why is this unresolvable by DNS? Will it always be unresolvable? Would something completely invalid be better? `not*a*valid*hostname`?

zmrow replied:
The helper only renders if the name is _unresolvable_ or resolves to `127.0.0.1 or ::1`.  I realize that `amazon.com` isn't a realistic hostname, but it should resolve mostly anywhere and that's the behavior I was after.

webern commented on sources/api/schnauzer/src/lib.rs:
```diff
@@ -131,6 +131,10 @@ pub fn build_template_registry() -> Result<handlebars::Handlebars<'static>> {
         Box::new(helpers::kube_reserve_memory),
     );
+    template_registry.register_helper(
+        "add_unresolvable_hostname",
```
Ugh, we now have several each of kabob case and snake case. Maybe we should add a comment that specifies what the preference is and that we messed it up.

zmrow replied:
:( I didn't even see the kabob cased ones.

I can add a comment requesting snake case since that's what we started with and the majority of them are snake case.

webern reviewed (approved):
Don't we need to add something to README for the new settings. Possibly a warning in k8s docs about kubelet not liking a hostname change?

zmrow commented:
^ Added the comments @webern asked for

webern commented on sources/api/schnauzer/src/lib.rs:
```diff
@@ -118,6 +118,7 @@ pub fn build_template_registry() -> Result<handlebars::Handlebars<'static>> {
     // but isn't provided in the data given to the renderer
     template_registry.set_strict_mode(true);
 
+    // Prefer snake case for heper names (we accidentally created a few with kabob case)
```
```suggestion
    // Prefer snake case for helper names (we accidentally created a few with kabob case)
```

zmrow replied:
Oops - good catch.  Fixed.

zmrow commented:
^ Fixed my misspelling. :)

bcressey commented on packages/release/release.spec:
```diff
@@ -135,14 +135,14 @@ install -d %{buildroot}%{_cross_templatedir}
 install -p -m 0644 %{S:200} %{buildroot}%{_cross_templatedir}/motd
 install -p -m 0644 %{S:201} %{buildroot}%{_cross_templatedir}/proxy-env
 install -p -m 0644 %{S:202} %{buildroot}%{_cross_templatedir}/hostname-env
+install -p -m 0644 %{S:10} %{buildroot}%{_cross_templatedir}/hosts
```
nit: this should be in the 200 range like other templates

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -98,6 +101,9 @@ const KUBE_RESERVE_3_CORES: f32 = KUBE_RESERVE_2_CORES + 5.0;
 const KUBE_RESERVE_4_CORES: f32 = KUBE_RESERVE_3_CORES + 5.0;
 const KUBE_RESERVE_ADDITIONAL: f32 = 2.5;
 
+const IPV4_LOCALHOST: &str = "127.0.0.1";
+const IPV6_LOCALHOST: &str = "::1";
```
You can make these `IpAddr` constants to avoid doing the conversion at runtime:
```suggestion
const IPV4_LOCALHOST: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
const IPV6_LOCALHOST: IpAddr = IpAddr::V6(Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 1));
```

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -869,6 +881,81 @@ pub fn kube_reserve_cpu(
+    if !hostname_resolveable {
+        let ipv4_entry = format!("{} {}", IPV4_LOCALHOST, hostname_str);
+        let ipv6_entry = format!("{} {}", IPV6_LOCALHOST, hostname_str);
+        let entries = format!("{}\n{}", ipv4_entry, ipv6_entry);
```
It might be better to ensure that a newline is written at the end of the file, rather than relying on handlebars - unless this results in an extra newline.

```suggestion
        let entries = format!("{}\n{}\n", ipv4_entry, ipv6_entry);
```

zmrow replied:
I'd prefer not to add an extra newline since handlebars seems to add one:
```
bash-5.0# cat -E /etc/hosts 
127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4$
::1 localhost localhost.localdomain localhost6 localhost6.localdomain6$
127.0.0.1 mrow-9-1$
::1 mrow-9-1$
```

zmrow replied:
Adding the extra newline in the template helper adds an additional line in the rendered template:
```
bash-5.0# cat -E /etc/hosts 
127.0.0.1 localhost localhost.localdomain localhost4 localhost4.localdomain4$
::1 localhost localhost.localdomain localhost6 localhost6.localdomain6$
127.0.0.1 mrow-9-1-newline$
::1 mrow-9-1-newline$
$
```

zmrow commented:
^ Addressed @bcressey 's comments

bcressey commented on sources/api/schnauzer/src/helpers.rs:
```diff
@@ -149,6 +154,12 @@ mod error {
             template: String,
         },
 
+        #[snafu(display("Invalid IP address '{}': {}", ip, source))]
+        IpFromString {
+            ip: String,
+            source: std::net::AddrParseError,
+        },
+
```
I don't think this error is used any more.

zmrow commented:
^ removed the unused error variant
